### PR TITLE
PWX-32229: added validate for storage and cloudstorage spec

### PR DESCRIPTION
**Pull request by svijaykumar-px**
**What this PR does / why we need it**: Added a Validate function to ensure than both storage and cloudstorage specs are not present at a cluster level or at the node level.

**Which issue(s) this PR fixes** : Closes PWX-32229

**Additional Notes** : Also changed a small part of the logic of portworx driver to copy cluster storage or cloud storage specs to the node specs only if there is no existing storage specs for the node.
Also added UTs to test the invalid specs

### CODE REVIEW

**Comment by zoxpx** [comment]
@svijaykumar-px -- did you test your change on the actual cluster (using local, and cloud storage)?  What was the effect when both `Storage` and `CloudStorage` were defined in the YAML-spec?

I expected the fix would detect the anomalies in the provided configuration (StorageCluster storage-segment), send the appropriate warnings/errors/events back to the user, and "do nothing" until this is corrected -- and your PR does not seem to be doing this.  (cc: @piyush-nimbalkar)

**Comment by svijaykumar-px** [comment]
> @svijaykumar-px -- did you test your change on the actual cluster (using local, and cloud storage)? What was the effect when both `Storage` and `CloudStorage` were defined in the YAML-spec?
> 
> I expected the fix would detect the anomalies in the provided configuration (StorageCluster storage-segment), send the appropriate warnings/errors/events back to the user, and "do nothing" until this is corrected -- and your PR does not seem to be doing this. (cc: @piyush-nimbalkar)

I tested the changes on an actual cluster with all the possible configurations.  The changes I made provides an error when there are both cloudstorage and storage specs at a cluster level or node level. It moves the storage cluster to a degraded stage and does not proceed till the specs are changed. So in essence It gives an error with the wrong specs and "does nothing" till it is corrected.

File: pkg/controller/storagecluster/controller_test.go [hunk 1]
```diff
@@ -1742,11 +1742,11 @@ func TestStoragePodGetsScheduledWithCustomNodeSpecs(t *testing.T) {
...
-			CloudStorage: &corev1.CloudStorageNodeSpec{
-				CloudStorageCommon: corev1.CloudStorageCommon{
-					DeviceSpecs: stringSlicePtr([]string{"type=dev1"}),
-				},
-			},
+			// CloudStorage: &corev1.CloudStorageNodeSpec{
+			// 	CloudStorageCommon: corev1.CloudStorageCommon{
+			// 		DeviceSpecs: stringSlicePtr([]string{"type=dev1"}),
+			// 	},
+			// },
```
Comment (zoxpx) [inline pkg/controller/storagecluster/controller_test.go:1749]: please remove commented code before committing

File: pkg/controller/storagecluster/controller_test.go [hunk 2]
```diff
@@ -1922,8 +1922,8 @@ func TestStoragePodGetsScheduledWithCustomNodeSpecs(t *testing.T) {
 		driver.EXPECT().GetStoragePodSpec(gomock.Any(), "k8s-node-1").
 			DoAndReturn(func(c *corev1.StorageCluster, _ string) (v1.PodSpec, error) {
 				require.Equal(t, cluster.Spec.Nodes[0].Storage, c.Spec.Storage)
-				require.Equal(t, cluster.Spec.Nodes[0].CloudStorage.CloudStorageCommon,
-					c.Spec.CloudStorage.CloudStorageCommon)
+				// require.Equal(t, cluster.Spec.Nodes[0].CloudStorage.CloudStorageCommon,
+				// c.Spec.CloudStorage.CloudStorageCommon)
```
Comment (zoxpx) [inline pkg/controller/storagecluster/controller_test.go:1926]: please remove commented code before committing

**Comment by zoxpx** [comment]
> I tested the changes on an actual cluster with all the possible configurations. The changes I made provides an error when there are both cloudstorage and storage specs at a cluster level or node level. It moves the storage cluster to a degraded stage and does not proceed till the specs are changed. So in essence It gives an error with the wrong specs and "does nothing" till it is corrected. @zoxpx Hopefully this answered your concerns

Thanks @svijaykumar-px -- that's what I wanted to hear.

It would have been helpful if you described your test and copy-pasted the result  (i.e. "added both storage + cloudStorage, STC switched to Degraded, displaying the following error ..") -- maybe next time.  :)

**Review by zoxpx — APPROVED** [review]
Thanks Swarupa -- some minor remarks below, please fix before comitting.

**Comment by svijaykumar-px** [comment]
> > I tested the changes on an actual cluster with all the possible configurations. The changes I made provides an error when there are both cloudstorage and storage specs at a cluster level or node level. It moves the storage cluster to a degraded stage and does not proceed till the specs are changed. So in essence It gives an error with the wrong specs and "does nothing" till it is corrected. @zoxpx Hopefully this answered your concerns
> 
> Thanks @svijaykumar-px -- that's what I wanted to hear.
> 
> It would have been helpful if you described your test and copy-pasted the result (i.e. "added both storage + cloudStorage, STC switched to Degraded, displaying the following error ..") -- maybe next time. :)

Oh okay sure, I will take note to add testing notes henceforth. Thank you!

**Pull request merged**